board.py

```diff
@@ -2,79 +2,170 @@
+        """Return True if position is within bounds of the game grid."""
+        return 0 <= x_pos - 1 <= self.width and 0 <= y_pos - 1 <= self.height
+
+    def update_board(self, snake_id_move_mapping):
+        """State transition for board. Applies moves to all live snakes.
+        * Grid is invalidated until the end of this method *
+        1. Move snakes
+        2. Snakes dying
+        3. Update world
+        4. Update grid"""
+
+        self.move_snakes(snake_id_move_mapping)
+        self.resolve_deaths()
+        self.update_board_after_move()
+        self.build_grid()
+
+    def move_snakes(self, snake_id_move_mapping):
+        """Apply moves to snakes."""
+        for snake_id, move in snake_id_move_mapping.item():
+            self.snakes[snake_id].apply_move_to_snake(move)
+
+    def resolve_deaths(self):
+        """Resolve deaths from starvation, wall collision, and snake collision, snakes that die the
+        current turn are moved from snakes to dead_snakes."""
+        starve_dead, starve_live = set(), set()
+        for snake in self.snakes:
+            (starve_dead, starve_live)[snake.health_points > 0].add(snake)
+
+        wall_dead, wall_live = set(), set()
+        for snake in self.snakes:
+            (wall_dead, wall_live)[self.cell_within_bounds(snake.head)].add(snake)
+
+        collision_dead, collision_live = set(), set()
+        cells_with_snakes_mapping = self._get_cells_with_snakes_mapping()
+        for snake in self.snakes:
+            snakes_at_cell = set(cells_with_snakes_mapping[snake.head])
+            no_collision = len(snakes_at_cell) == 1
+            collision_largest_head = self._is_collision_largest_head(snake, snakes_at_cell)
+            (collision_dead, collision_live)[no_collision or collision_largest_head].add(snake)
+
+        live_snakes = starve_live.intersection(wall_live).intersection(collision_live)
+        dead_snakes = starve_dead.union(wall_dead).union(collision_dead)
+
+        self.snakes = live_snakes
+        self.dead_snakes.append(list(dead_snakes))
+
+    def _get_cells_with_snakes_mapping(self):
+        """Get the mapping between cells and a list of snakes that occupy that cell."""
+        cells_with_snakes_mapping = {}
+        for snake in self.snakes:
+            for cell in snake.body:
+                if cell in cells_with_snakes_mapping:
+                    cells_with_snakes_mapping[cell].append(snake)
+                else:
+                    cells_with_snakes_mapping[cell] = [snake]
+
+    def _is_collision_largest_head(self, snake, snakes_at_cell):
+        """Return True if snake is the longest snake in the collision and that collision only
+        occured between snake heads."""
+        other_collided_snakes = snakes_at_cell.remove(snake)
+        for other_collided_snake in other_collided_snakes:
+            if snake.head != other_collided_snake.head:
+                return False
+        return len(snake.body) > max(other_collided_snakes,
```

Comment:
Implement a ``__len__`` fn in sanke then just use `len(snake)` and `max(snakes, key=len`